src/merkle/tiered_smt/mod.rs

```diff
@@ -0,0 +1,407 @@
+                }
+
+                if sibling_index.is_max_depth() {
+                    self.append_lower_leaf(sibling_index.into(), sibling)?;
+                    self.append_lower_leaf(index.into(), leaf)?;
+                } else {
+                    self.replace_upper_leaf(sibling_index.into(), sibling)?;
+                    self.storage.take_leaf_key(&node)?;
+                    self.replace_upper_leaf(index.into(), leaf)?;
+                }
+            }
+        }
+
+        Ok(self.root())
+    }
+
+    // HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Fetch a tree node, returning an empty sub-tree constant if absent.
+    fn get_node(&self, index: &NodeIndex) -> Result<Word, StorageError> {
+        let node = self
+            .storage
+            .get_node(index)?
+            .or_else(|| {
+                EMPTY_SUBTREES
+                    .get(index.depth() as usize)
+                    .copied()
+                    .map(Word::from)
+            })
+            .unwrap_or_default();
+        Ok(node)
+    }
+
+    /// Builds a list of leaves given its keys.
+    ///
+    /// The argument should be fetched from the storage as they correspondent values should exist
+    /// there.
+    fn siblings_keys_into_leaves<I>(&self, keys: I) -> Result<Vec<Leaf>, StorageError>
+    where
+        I: IntoIterator<Item = CanonicalWord>,
+    {
+        keys.into_iter()
+            .map(|k| {
+                self.storage.get_leaf_value(&k).map(|v| match v {
+                    Some(value) => Leaf::new(k, value),
+                    None => unreachable!(
+                        "a leaf key exists in the storage, but a corresponding value doesn't"
+                    ),
+                })
+            })
+            .collect()
+    }
+
+    /// Traverse both indexes until an empty or leaf node is found.
+    ///
+    /// The tree definition will not allow a lower tier to contain an internal node, so this
+    /// function will always yield either leaf or empty types.
+    fn traverse_until_empty_or_leaf(
+        &self,
+        leaf: &mut LeafIndex,
+        node: &mut NodeIndex,
+    ) -> Result<NodeType, StorageError> {
+        let mut current = self.get_type(node)?;
+        while current.is_internal() {
+            *node = match leaf.next() {
+                Some(node) => node,
+                None => unreachable!("a path always ends with either empty or leaf"),
+            };
+            current = self.get_type(node)?;
+        }
+        Ok(current)
+    }
+
+    /// Get a merkle path from the node to the root.
+    fn get_node_path(&self, mut index: NodeIndex) -> Result<MerklePath, StorageError> {
+        let mut path = Vec::with_capacity(index.depth() as usize);
+        while index.depth() > 0 {
+            let sibling = self.get_node(&index.sibling())?;
+            path.push(sibling);
+            index.reverse();
```

Comment:
nit: I find the name of `reverse()` a bit unintuitive. We are not really reversing the index but moving up the tree, right? Maybe `move_up()` or something like that would be a better name?

Reply:
I was trying to keep consistency with the opposite method, that is `traverse`. What about `backtrack`?